### ISSUE TITLE
I can't build the project

### ISSUE THREAD

**Issue by Limonado** [TheSHEEEP/godotdetour#1]
Hello, i'm making a game heavily dependant on AI and pathfinding, since the godot default navigation system ins't quite good, i searched and found this plugin, so i downloaded and installed all the components needed according to [https://docs.godotengine.org/en/3.2/development/compiling/compiling_for_windows.html](url) and started following the steps to build provided in the README.md, but when i try to build godotcdd the command prompt send this error:
```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
cl /Fosrc\detournavigationmesh.obj /c src\detournavigationmesh.cpp /TP /nologo -
W3 -GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARNINGS /DNDE
BUG /I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\include\core /
Igodot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\Include /I
recastnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Include /Irecastn
avigation\DetourTileCache\Include /Irecastnavigation\Recast\Include
cl /Fosrc\godotdetour.obj /c src\godotdetour.cpp /TP /nologo -W3 -GR -O2 -EHsc -
MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARNINGS /DNDEBUG /I. /Igodot-cp
p\godot_headers /Igodot-cpp\include /Igodot-cpp\include\core /Igodot-cpp\include
\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\Include /Irecastnavigation\D
etour\Include /Irecastnavigation\DetourCrowd\Include /Irecastnavigation\DetourTi
leCache\Include /Irecastnavigation\Recast\Include
detournavigationmesh.cpp
cl /Fosrc\util\chunkytrimesh.obj /c src\util\chunkytrimesh.cpp /TP /nologo -W3 -
GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARNINGS /DNDEBUG
/I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\include\core /Igod
ot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\Include /Ireca
stnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Include /Irecastnavig
ation\DetourTileCache\Include /Irecastnavigation\Recast\Include
cl /Fosrc\util\detourinputgeometry.obj /c src\util\detourinputgeometry.cpp /TP /
nologo -W3 -GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARNIN
GS /DNDEBUG /I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\includ
e\core /Igodot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\In
clude /Irecastnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Include /
Irecastnavigation\DetourTileCache\Include /Irecastnavigation\Recast\Include
godotdetour.cpp
chunkytrimesh.cpp
detourinputgeometry.cpp
c:\test\godot-cpp\include\core\Vector2.hpp(218): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(218): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(218): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
cl /Fosrc\util\godotdetourdebugdraw.obj /c src\util\godotdetourdebugdraw.cpp /TP
 /nologo -W3 -GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARN
INGS /DNDEBUG /I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\incl
ude\core /Igodot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\
Include /Irecastnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Include
 /Irecastnavigation\DetourTileCache\Include /Irecastnavigation\Recast\Include
c:\test\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
godotdetourdebugdraw.cpp
src\util\detourinputgeometry.cpp(288): warning C4244: 'initializing': conversion
 from 'int64_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(312): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(316): warning C4244: '=': conversion from 'int6
4_t' to 'unsigned char', possible loss of data
src\util\detourinputgeometry.cpp(317): warning C4244: '=': conversion from 'int6
4_t' to 'unsigned char', possible loss of data
src\util\detourinputgeometry.cpp(318): warning C4244: '=': conversion from 'int6
4_t' to 'unsigned short', possible loss of data
src\util\detourinputgeometry.cpp(319): warning C4244: '=': conversion from 'int6
4_t' to 'unsigned int', possible loss of data
src\util\detourinputgeometry.cpp(334): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(337): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(348): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(349): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(353): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(357): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(358): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(359): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(372): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(378): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\detourinputgeometry.cpp(388): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
c:\test\src\detournavigationmesh.h(18): warning C4099: 'rcConfig': type name fir
st seen using 'class' now seen using 'struct'
c:\test\src\detournavigationmesh.h(14): note: see declaration of 'rcConfig'
c:\test\src\detournavigationmesh.h(18): warning C4099: 'rcConfig': type name fir
st seen using 'class' now seen using 'struct'c:\test\godot-cpp\include\core\Vect
or2.hpp(218): warning C4244: 'argument': conversion from 'double' to 'real_t', p
ossible loss of data

c:\test\src\detournavigationmesh.h(14): note: see declaration of 'rcConfig'c:\te
st\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conversio
n from 'double' to 'real_t', possible loss of data

recastnavigation\Recast\Include\Recast.h(199): warning C4099: 'rcConfig': type n
ame first seen using 'class' now seen using 'struct'
recastnavigation\Recast\Include\Recast.h(198): note: see declaration of 'rcConfi
g'
recastnavigation\DetourCrowd\Include\DetourCrowd.h(117): warning C4099: 'dtCrowd
Agent': type name first seen using 'class' now seen using 'struct'
recastnavigation\DetourCrowd\Include\DetourCrowd.h(116): note: see declaration o
f 'dtCrowdAgent'
src\detournavigationmesh.cpp(230): error C2131: expression did not evaluate to a
 constant
src\detournavigationmesh.cpp(230): note: failure was caused by a read of a varia
ble outside its lifetime
src\detournavigationmesh.cpp(230): note: see usage of 'this'
src\detournavigationmesh.cpp(271): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(376): warning C4244: 'initializing': conversion fro
m 'int64_t' to 'int', possible loss of data
src\detournavigationmesh.cpp(381): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
cl /Fosrc\util\godotgeometryparser.obj /c src\util\godotgeometryparser.cpp /TP /
nologo -W3 -GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARNIN
GS /DNDEBUG /I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\includ
e\core /Igodot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\In
clude /Irecastnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Include /
Irecastnavigation\DetourTileCache\Include /Irecastnavigation\Recast\Include
src\detournavigationmesh.cpp(383): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(384): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(385): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(386): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(387): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(388): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(402): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(407): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(408): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(410): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(411): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(412): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(413): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(414): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(415): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(416): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of datagodotgeometryparser.cpp

src\detournavigationmesh.cpp(418): warning C4244: '=': conversion from 'int64_t'
 to 'int', possible loss of data
src\detournavigationmesh.cpp(501): warning C4244: 'initializing': conversion fro
m 'int64_t' to 'int', possible loss of data
src\detournavigationmesh.cpp(504): warning C4244: 'initializing': conversion fro
m 'int64_t' to 'int', possible loss of data
src\util\godotdetourdebugdraw.cpp(15): warning C4244: 'initializing': conversion
 from 'unsigned int' to 'float', possible loss of datasrc\detournavigationmesh.c
pp(604): warning C4018: '<': signed/unsigned mismatch

src\util\godotdetourdebugdraw.cpp(16): warning C4244: 'initializing': conversion
 from 'unsigned int' to 'float', possible loss of datasrc\detournavigationmesh.c
pp(607): warning C4018: '<': signed/unsigned mismatch

src\util\godotdetourdebugdraw.cpp(17): warning C4244: 'initializing': conversion
 from 'unsigned int' to 'float', possible loss of data
src\util\godotdetourdebugdraw.cpp(18): warning C4244: 'initializing': conversion
 from 'unsigned int' to 'float', possible loss of data
src\detournavigationmesh.cpp(613): warning C4018: '<': signed/unsigned mismatch
src\detournavigationmesh.cpp(616): warning C4018: '<': signed/unsigned mismatch
cl /Fosrc\util\meshdataaccumulator.obj /c src\util\meshdataaccumulator.cpp /TP /
nologo -W3 -GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WARNIN
GS /DNDEBUG /I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\includ
e\core /Igodot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtils\In
clude /Irecastnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Include /
Irecastnavigation\DetourTileCache\Include /Irecastnavigation\Recast\Include
src\detournavigationmesh.cpp(636): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(637): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(638): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(639): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(676): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(677): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
meshdataaccumulator.cpp
src\detournavigationmesh.cpp(686): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(687): warning C4244: 'initializing': conversion fro
m 'float' to 'int', possible loss of data
src\detournavigationmesh.cpp(962): error C2131: expression did not evaluate to a
 constant
src\detournavigationmesh.cpp(962): note: failure was caused by a read of a varia
ble outside its lifetime
src\detournavigationmesh.cpp(962): note: see usage of 'maxLayersToAdd'
src\detournavigationmesh.cpp(998): warning C4018: '<': signed/unsigned mismatch
src\detournavigationmesh.cpp(1009): warning C4018: '<': signed/unsigned mismatch

c:\test\godot-cpp\include\core\Vector2.hpp(218): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
cl /Fosrc\util\navigationmeshhelpers.obj /c src\util\navigationmeshhelpers.cpp /
TP /nologo -W3 -GR -O2 -EHsc -MD /DWIN32 /D_WIN32 /D_WINDOWS /D_CRT_SECURE_NO_WA
RNINGS /DNDEBUG /I. /Igodot-cpp\godot_headers /Igodot-cpp\include /Igodot-cpp\in
clude\core /Igodot-cpp\include\gen /Isrc /Isrc\util /Irecastnavigation\DebugUtil
s\Include /Irecastnavigation\Detour\Include /Irecastnavigation\DetourCrowd\Inclu
de /Irecastnavigation\DetourTileCache\Include /Irecastnavigation\Recast\Include
scons: *** [src\detournavigationmesh.obj] Error 2
navigationmeshhelpers.cpp
c:\test\godot-cpp\include\core\Vector2.hpp(218): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
src\util\godotgeometryparser.cpp(58): warning C4244: '=': conversion from 'int64
_t' to 'int', possible loss of data
src\util\godotgeometryparser.cpp(61): warning C4244: '=': conversion from 'int64
_t' to 'int', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(218): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
c:\test\godot-cpp\include\core\Vector2.hpp(217): warning C4244: 'argument': conv
ersion from 'double' to 'real_t', possible loss of data
src\util\godotgeometryparser.cpp(89): warning C4244: '=': conversion from 'int64
_t' to 'int', possible loss of data
src\util\godotgeometryparser.cpp(92): warning C4244: '=': conversion from 'int64
_t' to 'int', possible loss of data
src\util\meshdataaccumulator.cpp(25): warning C4018: '<': signed/unsigned mismat
ch
src\util\meshdataaccumulator.cpp(71): warning C4018: '<': signed/unsigned mismat
ch
src\util\meshdataaccumulator.cpp(78): warning C4018: '<': signed/unsigned mismat
ch
src\util\meshdataaccumulator.cpp(85): warning C4018: '<': signed/unsigned mismat
ch
src\util\meshdataaccumulator.cpp(95): warning C4244: 'initializing': conversion
from 'int64_t' to 'int', possible loss of data
src\util\meshdataaccumulator.cpp(101): warning C4244: 'initializing': conversion
 from 'int64_t' to 'int', possible loss of data
src\util\meshdataaccumulator.cpp(109): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
src\util\meshdataaccumulator.cpp(113): warning C4244: '=': conversion from 'int6
4_t' to '_Ty', possible loss of data
        with
        [
            _Ty=int
        ]
src\util\meshdataaccumulator.cpp(117): warning C4244: '=': conversion from 'int6
4_t' to 'int', possible loss of data
scons: building terminated because of errors.
```

I'm using windows 8.1 with visual studio 2017, scons 3.1.2 and python 3.7.7, i never worked with building or c++ so i don't know if i've done something wrong 

I tried with visual studio 2019 too but had the same error

**Comment by TheSHEEEP** [comment]
The error c2131 kinda says it all. MSVC is known to not be the friendliest compiler, to put it mildly.
These code pieces probably need to be changed to compile on Windows. Should be easy enough to do, but I don't know when I'll get around to launch Windows again in the next few days.

For the moment, I'd suggest changing the offending lines to be MSVC compatible. I'll also gladly take a pull request or patch.

**Comment by TheSHEEEP** [comment]
Just to reiterate, I'll definitely fix this (I need this module to work on Windows as well), but it might be a few days before I get to it.

**Comment by Limonado** [comment]
> The error c2131 kinda says it all. MSVC is known to not be the friendliest compiler, to put it mildly.
> These code pieces probably need to be changed to compile on Windows. Should be easy enough to do, but I don't know when I'll get around to launch Windows again in the next few days.
> 
> For the moment, I'd suggest changing the offending lines to be MSVC compatible. I'll also gladly take a pull request or patch.

Thanks for responding, i fixed the errors and got it building correctly, but when i start the demo and try to place an agent or a obstacle it crashes and closes without showing any errors, probably i messed up something while fixing but i don't know for sure

The only things i can do in the demo is create the navmesh, enable debug drawing and mark place as water, the other options results in a crash or don't do nothing

**Comment by TheSHEEEP** [comment]
I finally got around to my Windows machine and was able to reproduce everything you wrote, including the crash. Without errors - wtf?!

Either way, a bit of going wild with logging later, I found the culprit, though I don't have a solution yet:
https://github.com/TheSHEEEP/godotdetour/blob/master/src/detournavigation.cpp#L394

On Linux, this works fine producing a valid instance of the agent. On Windows, it produces an invalid Null instance. I don't get it, either - I'll ask around on the Discord.
